### ISSUE TITLE
Cannot take gradient of function using linspace 

### ISSUE THREAD

**Issue by ronan-keane** [google/jax#1571]
I've found I will get an exception "Tracer can't be used with raw numpy function" when I use linspace where one of the arguments to linspace is a parameter which needs to be differentiated with respect to. Here is a MWE that shows the problem 

```
import jax.numpy as jnp
import jax 

def testlinspace(x, *args): #does not work 
    out = jnp.linspace(x[0],x[1],50) #linspace depends on parameters 
    
    out = x[2]*jnp.square(out)
    out = jnp.sum(out)
    
    return out 

def testlinspace2(x, *args): #works 
    out = jnp.linspace(1.,50.,50) #linspace is just some array of numbers 
    
    out = x[2]*jnp.square(out)+x[0]*x[1]
    out = jnp.sum(out)
    
    return out 

x = [2.,3.,1.5]
testobj = testlinspace(x)
testgrad = jax.grad(testlinspace)
testobj2 = testlinspace2(x)
testgrad2 = jax.grad(testlinspace2)

print(testgrad(x)) #throws exception, note gradient using finite differences is DeviceArray([174.48979293, 200.51020897, 316.83673001])
print(testgrad2(x)) #works, gives DeviceArray([  149.99968698,    99.99930626, 42924.99943404])
```

So the problem is you can't do something like jax.numpy.linspace(x[0],10,50) because the linspace is depending on the parameters 'x' being reverse-mode differentiated.

**Comment by levskaya** [comment]
Yeah, we seem to just be wrapping the original numpy implementation.  The below is a draft of a native differentiable form that mirrors the numpy api close to machine precision for an extensive set of all the linspace arguments that I could think to test.

```python
import numpy as onp
from jax import lax
import jax.numpy as np
def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None, 
             axis=0):
  """Implementation of linspace differentiable w.r.t. start and stop args."""
  lax._check_user_dtype_supported(dtype, "linspace")
  dtype = onp.float32 if dtype is None else dtype
  bounds_shape = list(lax.broadcast_shapes(onp.shape(start), onp.shape(stop)))
  axis = len(bounds_shape) if axis == -1 else axis
  bounds_shape.insert(axis, 1)
  iota_shape = [1,] * len(bounds_shape)
  iota_shape[axis] = num
  delta = (stop - start) / num
  if endpoint:
    delta *= num / (num - 1)
  out = (np.reshape(start, bounds_shape) + 
         np.reshape(lax.iota(dtype, num), iota_shape) * 
         np.reshape(delta, bounds_shape))
  if retstep:
    return np.array(out, dtype=dtype), delta
  else:
    return np.array(out, dtype=dtype)
```

you can see it working with your gradient test in the public colab notebook:
https://colab.research.google.com/drive/1ARqNihdk53NOxbdRxzPv4NmrtwiY6qZ9 

If this looks good to others we could upstream it to replace the current wrapper implementation.

**Comment by ronan-keane** [comment]
I can't see the notebook

**Comment by levskaya** [comment]
My apologies, this should be a public link (edited above as well): 
https://colab.research.google.com/drive/1bdyUypR-S2stCe2H_LyrEBB4sSFZ9ssL

**Comment by ronan-keane** [comment]
Sorry it took me awhile to reply. I looked at your solution and verified it worked for my MWE. I tested this for a more complex example as well and it also worked for me there. So in my opinion this should replace the existing jax linspace.

Please let me know if you want me to close this issue.
